### PR TITLE
Add ConversationDirectoryType protocol

### PR DESCRIPTION
**Pull request by typfel**
## What's new in this PR?

Introduce the `ConversationDirectoryType` protocol which provides a new way of accessing the conversations which should be displayed in the UI.

For now this is build on top of the existing `ZMConversationListDirectory` and `ZMConversationList`.

## TODO

- [x] Add tests

### CODE REVIEW

File: Source/Model/Conversation/ZMConversation+Predicates.swift [hunk 1]
```diff
@@ -92,7 +92,44 @@ extension ZMConversation {
...
+        
+        return NSCompoundPredicate(andPredicateWithSubpredicates: [predicateForConversationsExcludingArchived(), groupConversationPredicate])
+    }
+    
+    class func predicateForUnconnectedConversations() -> NSPredicate {
+        return NSPredicate(format: "\(ZMConversationConversationTypeKey) == \(ZMConversationType.connection.rawValue)")
+    }
+    
+    class func predicateForOneToOneConversation() -> NSPredicate {
+        return NSPredicate(format: "\(ZMConversationConversationTypeKey) == \(ZMConversationType.group.rawValue)")
```
Comment (billypchan) [inline Source/Model/Conversation/ZMConversation+Predicates.swift]: Should it be `ZMConversationType.oneOnOne.rawValue` ?

Comment (typfel) [reply Source/Model/Conversation/ZMConversation+Predicates.swift]: I'll also add a test for this.

File: Source/ConversationList/ConversationDirectory.swift [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+//
+
+import Foundation
+
+public enum ConversationListType {
+    case archived, unarchived, pending, contacts, groups
+}
+
+public struct ConversationDirectoryChangeInfo {
+    var reloaded: Bool
```
Comment (billypchan) [inline Source/ConversationList/ConversationDirectory.swift]:
```suggestion
    public var reloaded: Bool
```

Comment (billypchan) [reply Source/ConversationList/ConversationDirectory.swift]: Is it possible to provide the updated folder here?

Comment (typfel) [reply Source/ConversationList/ConversationDirectory.swift]: You means something like `var updatedLists: [ConversationListType]`?

File: Source/ConversationList/ConversationDirectory.swift [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        }
+    }
+    
+    public var folders: [FolderType] {
+        return []
+    }
+    
+    public func addObserver(_ observer: ConversationDirectoryObserver) -> Any {
+        let observerProxy = ConversationListObserverProxy(observer: observer)
+        let token = ConversationListChangeInfo.add(observer: observerProxy, managedObjectContext: nil!)
```
Comment (billypchan) [inline Source/ConversationList/ConversationDirectory.swift]: `managedObjectContext` is not nullable here.

Comment (typfel) [reply Source/ConversationList/ConversationDirectory.swift]: This should have been a todo

**Pull request merged**